### ISSUE TITLE
Compatibility with WARUN

### ISSUE THREAD

**Issue by PorscheMonty** [chocolate-doom/chocolate-doom#621]
There's this app called WARUN.EXE (WadAuthor run map utility) that comes bundled with WadAuthor v1.30, and is used of WAUTHOR.EXE itself for test-running the map you're working on..

Little is known about it's inner works, but I thought that having a valid doom2 installation and basically replacing vanilla doom2.exe with a like-named Chocolate Doom executable, would have been enough for WARUN to launch a game, but that doesn't seem to work at all (closes back to desktop after a quick black windowed screen)

Yes, I know there's a way around it, for WadAuthor makes provision for customized third-party application launching, but unless there's some kind of twisted checksum thing getting in the way or what have you, it stands to reason WARUN should work with Chocolate Doom as well, and that's the motivation behind this report.

Regards.

**Comment by haleyjd** [comment]
I disassembled the file, it is a 16-bit DOS program built with Microsoft C 6.0 or Visual C 1.5. It uses the DOS library function execv() to execute the program. Windows, even where 16-bit program execution is allowed, does not allow 16-bit processes to spawn 32- or 64-bit Windows processes.

**Comment by PorscheMonty** [comment]
Thank you for taking the time to dive into this. Good to know everything's fine on CD's end.

**Comment by fragglet** [comment]
WadAuthor itself (wauthor.exe) seems to be a 32-bit executable. So I guess it should be possible to replace it with a simple 32-bit executable that runs chocolate-doom.exe instead. According to the command line, the arguments it takes are a directory (to cd to?), executable name, and any further arguments are passed through to Doom.

There's an interesting side point here though. It looks like we can solve this for WadAuthor trivially by replacing warun.exe, but that won't work for other editors and utilities, especially those that are DOS-based. It would be nice to have a "trampoline" DOS binary for compatibility that would allow such utilities to launch Chocolate Doom. Based on what Quasar says though, I'm not sure how such a trampoline could actually work.

**Comment by haleyjd** [comment]
Anything I can think of that might theoretically work under NTVDM (which might allow for instance a 16-bit Windows app to access the CreateProcess API) would definitely not work under DosBox, which is how 99% of Win64 users have to run DOS stuff. I think it's probably an exercise in futility.

**Comment by fragglet** [comment]
Yes, ideally the dosbox use case would be supported too. That's increasingly important nowadays. At first appearance that sounds impossible but it isn't necessarily so.

DOSbox has builtin support for modem emulation which lets you "dial" arbitrary TCP ports. So theoretically you could run an agent on your desktop that waited for TCP connections on a known port. The DOS utilities run a fake doom2.exe that connects to the agent and launches the game.

All ridiculously complicated of course but could be a fun hack. And bonus points if -net / -statcopy / -control can be made to work too.

**Comment by PorscheMonty** [comment]
A simple WARUN replacement exe that works with Chocolate Doom sounds like good a deal, WadAuthor having long become enough a staple mapping utility.

Anything beyond that, I'd say it may be fun, but not all that much necessary.

**Comment by haleyjd** [comment]
I didn't consider DosBox's networking capability :)  It *is* a long way around to accomplish something simple, but it's not conceptually all that different from a pipe with an underlying network transport for RPC either ;)
